### DISCUSSION TITLE
[@types/estree] Missing decorators inside node.body.body, and missing decorators inside Pattern

### DISCUSSION THREAD

**Discussion by zhaoyiming0803** [DefinitelyTyped/DefinitelyTyped#58605] [Issues with a @types package]
I am writing an eslint plugin, the code is as follows

``` typescript
import { MethodDefinition, ClassDeclaration } from 'estree'

export const useClassValidatorToDto = {
	meta: {
		messages
	},
	create(context) {
		return {
			ClassDeclaration(node: ClassDeclaration) {
				// The type of nodeItem should to be MethodDefinition or PropertyDefinition
                                 // but ts error: Property 'decorators' does not exist on type 'MethodDefinition | PropertyDefinition'.
                                 // in fact, the runtime has this value(nodeItem.decorators)
				node.body.body.forEach((nodeItem: any) => {
					if (!nodeItem.decorators.length) {
						// ...
					}
				})
			},
			MethodDefinition: (node: MethodDefinition) => {
				// The type of bodyParam should to be Pattern
				const bodyParam: any = node.value.params.find(p => getDecoratorByName(p, 'Body'))
			}
		}
	}
}
```

I can't find the related type definitions, so use `any`. but i don't want to.

[a comment by an automated account is not shown]

**Comment by RReverser** [comment]
It's because decorators are not a finalized spec, and, correspondingly, not part of ESTree.

**Reply by zhaoyiming0803** [reply]
> It's because decorators are not a finalized spec, and, correspondingly, not part of ESTree.

Thanks! Is there any other better solution? Except `any`

**Reply by RReverser** [reply]
You can define your own extension in project-specific `.d.ts` file, or look around in case someone already implemented those. TypeScript supports merging interfaces from different files, so that should work.

**Reply by zhaoyiming0803** [reply]
Thank you very much. It seems that this is the only way to do it.

**Reply by zhaoyiming0803** [reply]
> You can define your own extension in project-specific `.d.ts` file, or look around in case someone already implemented those. TypeScript supports merging interfaces from different files, so that should work.

I has created a file named `src/types/estree.d.ts`: 

``` typescript
import {
  Pattern
} from 'estree'

declare module 'estree' {
  export interface ICallee {
    name: string
    type: 'Identifier'
    [name: string]: any
  }
   interface IMethodDefinition extends MethodDefinition {
    value: IFunctionExpression
    decorators?: IDecorator[]
  }
  // ...
}
```
but this file does not exist in the compiled product, my tsconfig.json is as follows:

``` json
{
  "compilerOptions": {
    "target": "es2015",
    "baseUrl": ".",
    "module": "commonjs",
    "moduleResolution": "node",
    "importHelpers": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "declaration": true,
    "outDir": "dist",
    "allowJs": false,
    "removeComments": false,
    "strict": true,
    "noUnusedLocals": true,
    "noImplicitAny": false,
    "types": ["node"],
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "include": [
    "src/**/*.ts"
  ],
  "exclude": [
    "node_modules/**"
  ]
}
```

`dist/index.d.ts` has error:

``` typescript
export declare const rules: {
    forbid_body_parameters: {
        meta: {
            messages: {
                invalidArgument: string;
            };
        };
        create(context: any): {
            MethodDefinition: (node: import("estree").IMethodDefinition) => void;
        };
    };
}
```
IMethodDefinition is not existed in estree......has no exported member 'IMethodDefinition'

**Reply by RReverser** [reply]
1) I'm not sure why that doesn't work. Maybe try adding explicit filepath to `include`? Can't help there much otherwise, sorry.

2) What I meant is not create a new interface, but literally extend existing interfaces by redeclaring them; TypeScript will take care of merging them for you. That is, you can just write your extends like:

```ts
declare module 'estree' {
   interface MethodDefinition {
    decorators?: Decorator[]; // this will be added to all existing instances of `MethodDefinition`
  }
  // ...
}
```

Then you can use normal `estree` types everywhere, and they will have the added properties.

**Reply by zhaoyiming0803** [reply]
> 

Yes, you are right, i have solved the problem. Thank you !
